### PR TITLE
Add plugin version, Google Merchant Center account ID, and Google Ads account ID to all frontend tracking events

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

Project thread: pcTzPl-20H-p2

This PR implements the JS part of the project. The target is to attach the following properties when sending all JS tracking events:

- Plugin Version
- Google Merchant Center account ID if connected
- Google Ads account ID if connected

To achieve the above target, this PR:

- On the PHP side, add base tracking properties to the inline script of outputting the global JS variable `glaData`.
- Add a general state to this extension's wp-data store for accessing the prefetched data (`glaData`) so that the account IDs in frontend cache can be synced after connecting/disconnecting an account.
- To reduce confusion with the `recordEvent` function imported from `@woocommerce/tracks`
   - Rename the **js/src/utils/recordEvent.js** file to **tracks.js**.
   - In the **tracks.js**, add new functions to wrap the `recordEvent` and `queueRecordEvent` to attach the base tracking properties.
   - Replace all `recordEvent` calls with `recordGlaEvent`.
   - Replace all `queueRecordEvent` calls with `queueRecordGlaEvent`.
- Update tracking README.

💡 The `wcadmin_page_view` event is sent from Woo core, it will be handled by another PR

#### 📷 Base tracking properties

image

#### 📹 Attach Google Merchant Center account ID if connected

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/e4d6180f-de45-440a-b593-1b3702fd850e

#### 📹 Attach Google Ads account ID if connected

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/ccb4371e-62f8-4f4f-9913-f0105fe1d0cd

#### Queue a tracking event

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/a5eafe7b-a3d1-4efe-bbcf-a15811e8d94a

### Detailed test instructions:

💡 Apart from the `wcadmin_page_view` event, all tracking events should have the base properties.

1. Enable tracking debug tool
   - Open the Console tab of the browser DevTool.
   - Change the filter level to "All levels".
   - Run `localStorage.setItem( 'debug', 'wc-admin:*' )` and refresh the page to make it work.
   - It should be able to view tracking via the Console tab. Ref:
      
2. Disconnect Google Merchant Center and Google Ads accounts if already connected.
3. Go to the first step of onboarding flow.
4. Check if all tracking events have the `gla_version` property.
5. Check if all tracking events don't have the `gla_mc_id` property before connecting a GMC account.
6. Check if all tracking events have the `gla_mc_id` property and correct ID value after a GMC account is connected.
7. Proceed to the last step of onboarding flow.
8. Check if all tracking events don't have the `gla_ads_id` property before connecting a Google Ads account.
9. Check if all tracking events have the `gla_ads_id` property and correct ID value after a Google Ads account is connected.
10. Go to the Product Feed page.
11. Enable the "Preserve log" option in the Console tab.
12. Click one of the "Edit" links in the Product Feed table.
13. Wait for the product editing page to load.
14. Check if the queued tracking event `gla_edit_product_click` has the base properties.
   
15. Randomly operate the UI to trigger more tracking events to see if all have the base properties.
16. Global search the codebase to check if any `recordEvent` or `queueRecordEvent` call is missed to replace.

### Changelog entry

> Add - Include plugin version, Google Merchant Center account ID, and Google Ads account ID in all frontend tracking events.

### CODE REVIEW

**Comment by mikkamp** [comment]
@eason9487 Thanks for the implementation. Just had a follow up question about the connect account tracks.

For example if I click the button to connect account `123456789`, what event properties will get sent with `gla_ads_account_connect_button_click`? Will it already populate `gla_ads_id` or will the property only be updated after this event has been recorded?
I assume it will only be populated after, if that's the case then what kind of effort would it be to include it as a property? Doesn't have to be part of the base properties yet, could just be a property passed to recordGlaEvent.

**Comment by eason9487** [comment]
Hi @mikkamp 

> For example if I click the button to connect account `123456789`, what event properties will get sent with `gla_ads_account_connect_button_click`? Will it already populate `gla_ads_id` or will the property only be updated after this event has been recorded?

In this case, its event properties will only have `gla_version` and `gla_mc_id`, because at the moment the button is clicked, the selected Google Ads account is not yet connected. The `gla_ads_id` will be only included in subsequent events when an account is connected.

> I assume it will only be populated after, if that's the case then what kind of effort would it be to include it as a property? Doesn't have to be part of the base properties yet, could just be a property passed to recordGlaEvent.

If I didn't miss anything, there should be two related events:
1. `gla_mc_account_connect_button_click`
   https://github.com/woocommerce/google-listings-and-ads/blob/7e1c9301b6495c90c0eb0626f9484ec402fcfe2a/js/src/components/google-mc-account-card/connect-mc/index.js#L109-L117
2. `gla_ads_account_connect_button_click`
   https://github.com/woocommerce/google-listings-and-ads/blob/7e1c9301b6495c90c0eb0626f9484ec402fcfe2a/js/src/components/google-ads-account-card/connect-ads/index.js#L126-L133

, and they have a very similar code code structure for connecting the account.

Take the first one as an example, the minimum code change will be:

```diff
--- a/js/src/components/google-mc-account-card/connect-mc/index.js
+++ b/js/src/components/google-mc-account-card/connect-mc/index.js

@@ -28,2 +28,3 @@ import './index.scss';
  * @event gla_mc_account_connect_button_click
+ * @property {number} id The account id to be connected.
  */

@@ -113,2 +114,3 @@ const ConnectMC = () => {
   eventName="gla_mc_account_connect_button_click"
+  eventProps={ { id: Number( value ) } }
   onClick={ handleConnectMC }
```

The required effort should be pretty small.

**Edit**: Having verified that the above code change was correct, all the necessary practical work had been done as well, so I opened PR #2201 in passing.

**Comment by mikkamp** [comment]
Thanks for the clear explanation and the follow up PR. That resolves all my requests.

File: js/src/utils/tracks.js [hunk 1]
```diff
@@ -23,6 +30,48 @@ import { recordEvent } from '@woocommerce/tracks';
  * @property {string} direction Direction of page to be changed. `("next" | "previous")`
  */
 
+function prepareEventProperties( eventProperties ) {
```
Comment (puntope) [inline js/src/utils/tracks.js:33]: Missed docs

**Comment by puntope** [comment]
⚠️  `wcadmin_gla_setup_mc` is not showing GLA Version


https://github.com/woocommerce/google-listings-and-ads/assets/5908855/cc49186d-3de7-4ccf-92c4-245c7b0daf70

**Comment by puntope** [comment]
Not introduced in this PR. But when I was testing the Setup Ads Faq, I noticed it added the context as `undefined`

I created an Issue for this #2211

**Review by puntope — COMMENTED** [review]
Thanks for working on this. I left 2 comments in this PR to solve before approval

File: js/src/utils/tracks.js (hunk 1 above)
Comment (eason9487) [reply js/src/utils/tracks.js:33]: The JSDoc will be added by e14d03d63ceef338b218a7719f5db5e51073ec23 in #2200

**Comment by eason9487** [comment]
Hi @puntope, thanks for the review. Could you help with a new round of code reviews?

> ⚠️ `wcadmin_gla_setup_mc` is not showing GLA Version

The logged JS object may not be expandable after leaving the page:



I used Network requests to inspect the event. The `gla_version` property was included:

**Review by puntope — APPROVED** [review]
Thanks for clarifying the issue with the event props. This can be merged.

**Pull request merged**